File: frontend/src/types/question.ts
export interface QuestionBase {
  question_number: number
  question_text: string
  option_a: string
  option_b: string
  option_c: string
  option_d: string
  correct_answer: string
}

export interface QuestionCreate extends QuestionBase {
  chapter_id: number
}

export interface QuestionBulkCreate {
  chapter_id: number
  questions: QuestionBase[]
}

export interface QuestionUpdate {
  question_text?: string
  option_a?: string
  option_b?: string
  option_c?: string
  option_d?: string
  correct_answer?: string
}

export interface QuestionResponse extends QuestionBase {
  id: number
  chapter_id: number
  status: 'NEW' | 'MASTERED' | 'REVIEW' | 'ALMOST_FORGOT' | 'ERROR'
  times_attempted: number
  times_correct: number
  times_wrong: number
  created_at: string
  updated_at: string | null
}

export interface QuestionListResponse {
  data: QuestionResponse[]
  total: number
  page: number
  page_size: number
  message: string
}

export interface BulkSaveResponse {
  saved_count: number
  failed_count: number
  questions: QuestionResponse[]
  message: string
}

export type QuestionStatus = 'NEW' | 'MASTERED' | 'REVIEW' | 'ALMOST_FORGOT' | 'ERROR'

export interface QuestionStatusUpdate {
  status: QuestionStatus
}

export interface StatusUpdateResponse {
  question_id: number
  status: string
  message: string
}

export interface QuestionBulkStatusUpdate {
  updates: Array<{
    question_id: number
    status: QuestionStatus
  }>
}

export interface BulkStatusUpdateResponse {
  updated_count: number
  failed_count: number
  updates: StatusUpdateResponse[]
  message: string
}

File: backend/app/schemas/question.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List
from datetime import datetime


class QuestionBase(BaseModel):
    question_number: int
    question_text: str = Field(..., min_length=1)
    option_a: str = Field(..., min_length=1)
    option_b: str = Field(..., min_length=1)
    option_c: str = Field(..., min_length=1)
    option_d: str = Field(..., min_length=1)
    correct_answer: str = Field(..., pattern="^[A-D]$")
    explanation: Optional[str] = None

    @field_validator('question_text', 'option_a', 'option_b', 'option_c', 'option_d')
    @classmethod
    def validate_not_empty(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError('Field cannot be empty or whitespace only')
        return v

    @field_validator('correct_answer')
    @classmethod
    def validate_answer(cls, v: str) -> str:
        v = v.upper().strip()
        if v not in ['A', 'B', 'C', 'D']:
            raise ValueError('Correct answer must be A, B, C, or D')
        return v


class QuestionCreate(QuestionBase):
    chapter_id: int


class QuestionBulkCreate(BaseModel):
    """Schema for bulk creating questions"""
    chapter_id: int
    questions: List[QuestionBase]


class QuestionUpdate(BaseModel):
    question_text: Optional[str] = None
    option_a: Optional[str] = None
    option_b: Optional[str] = None
    option_c: Optional[str] = None
    option_d: Optional[str] = None
    correct_answer: Optional[str] = None


class QuestionStatusUpdate(BaseModel):
    """Schema for updating question status"""
    status: str = Field(..., pattern="^(NEW|MASTERED|REVIEW|ALMOST_FORGOT|ERROR)$")
    
    @field_validator('status')
    @classmethod
    def validate_status(cls, v: str) -> str:
        v = v.upper().strip()
        valid_statuses = ['NEW', 'MASTERED', 'REVIEW', 'ALMOST_FORGOT', 'ERROR']
        if v not in valid_statuses:
            raise ValueError(f'Status must be one of: {", ".join(valid_statuses)}')
        return v


class QuestionBulkStatusUpdate(BaseModel):
    """Schema for bulk updating question statuses"""
    updates: List[dict] = Field(..., description="List of {question_id, status} objects")
    
    @field_validator('updates')
    @classmethod
    def validate_updates(cls, v: List[dict]) -> List[dict]:
        if not v:
            raise ValueError('Updates list cannot be empty')
        
        valid_statuses = ['NEW', 'MASTERED', 'REVIEW', 'ALMOST_FORGOT', 'ERROR']
        
        for update in v:
            if 'question_id' not in update or 'status' not in update:
                raise ValueError('Each update must have question_id and status')
            
            if not isinstance(update['question_id'], int):
                raise ValueError('question_id must be an integer')
            
            status = update['status'].upper().strip()
            if status not in valid_statuses:
                raise ValueError(f'Status must be one of: {", ".join(valid_statuses)}')
            
            update['status'] = status
        
        return v


class QuestionResponse(QuestionBase):
    id: int
    chapter_id: int
    status: str
    times_attempted: int
    times_correct: int
    times_wrong: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class QuestionListResponse(BaseModel):
    data: List[QuestionResponse]
    total: int
    page: int
    page_size: int
    message: str = "Questions retrieved successfully"


class BulkSaveResponse(BaseModel):
    """Response for bulk save operation"""
    saved_count: int
    failed_count: int
    questions: List[QuestionResponse]
    message: str


class StatusUpdateResponse(BaseModel):
    """Response for status update operation"""
    question_id: int
    status: str
    message: str


class BulkStatusUpdateResponse(BaseModel):
    """Response for bulk status update operation"""
    updated_count: int
    failed_count: int
    updates: List[StatusUpdateResponse]
    message: str
